File: backend/app/schemas/setlist.py
from pydantic import BaseModel
from uuid import UUID
from datetime import date


class SongEntry(BaseModel):
    # 셋리스트 곡 하나
    name: str
    encore: bool = False
    # 페스티벌처럼 아티스트가 여럿인 공연에서, 이 곡이 누구 예상/실제 셋리스트
    # 소속인지 표시. 단독 공연이면 그냥 None(불필요).
    artist: str | None = None
    # 곡 출처 - "representative"면 과거 셋리가 없어 대신 채운 대표곡(화면에서 예상 셋리와
    # 구분 표시), None이면 Setlist.fm 셋리 기반
    source: str | None = None


class SetlistEditRequest(BaseModel):
    # 실제 셋리스트 유저 수정 요청
    songs: list[SongEntry]


class FetchSetlistRequest(BaseModel):
    # Setlist.fm ID로 셋리스트 가져와 저장 요청
    setlistfm_id: str


class ArtistSetlistStatus(BaseModel):
    # 실제 셋리가 빈 아티스트의 상태(앱 빈 화면 문구용) - searching(아직 안 찾아봄)/searched(찾았지만
    # 없음)/unresolved(mbid 없어 누구인지 확정 못함)/not_artist("아티스트가 아니에요"로 설정됨)
    artist: str
    state: str
    name: str | None = None
    # 누구로 찾았는지 알아보게 붙이는 대표곡 1곡(searching/searched만)
    top_song: str | None = None


class RealSetlistResponse(BaseModel):
    model_config = {"from_attributes": True}

    # 실제 셋리스트 조회 응답. row가 없어도(Setlist.fm 매칭 실패 등) id=None/songs=[]로
    # 200을 반환할 수 있음 - artist_names는 그 경우에도 채워져서 프론트가 아티스트별
    # 안내 문구를 보여줄 수 있음.
    id: UUID | None = None
    concert_id: UUID
    performance_date: date
    setlistfm_id: str | None
    songs: list[SongEntry]
    is_user_edited: bool
    edited_user_nickname: str | None
    artist_names: list[str] = []
    artist_statuses: list[ArtistSetlistStatus] = []


class SetlistFmCandidate(BaseModel):
    # Setlist.fm 검색 결과 후보 항목
    setlistfm_id: str
    event_date: str
    artist_name: str
    venue_name: str
    city_name: str
    song_count: int
    songs: list[SongEntry]
    url: str


class PreSetlistResponse(BaseModel):
    model_config = {"from_attributes": True}

    # 예상 셋리스트 조회 응답 - id가 None이면 아직 row가 없음(빈 songs + artist_names만)
    id: UUID | None = None
    concert_id: UUID
    setlistfm_id: str | None
    songs: list[SongEntry]
    is_user_edited: bool
    edited_user_nickname: str | None
    # RealSetlistResponse와 동일 패턴 - 프론트가 단독 공연에서 song.artist가
    # 비어있는 곡을 음악앱 검색으로 연결할 때 폴백 아티스트로 씀.
    artist_names: list[str] = []


class ArtistCandidate(BaseModel):
    # 앵커 후보(공연 아티스트 이름으로 찾은 iTunes 아티스트) - 동명이인은 장르/대표곡으로 구분
    itunes_artist_id: str
    artist_name: str
    genre: str | None = None
    top_songs: list[str] = []
    artwork_url: str | None = None
    # 공연 아티스트 이름과 정확히 같은지(아니면 협업/편집 앨범 등 비슷한 이름)
    exact_match: bool = False


class ArtistAnchorCandidate(BaseModel):
    # 앵커 후보(iTunes 곡 검색 결과) - 유저가 고르면 이 곡의 아티스트로 확정
    itunes_artist_id: str
    artist_name: str
    track_name: str
    album_name: str | None = None
    artwork_url: str | None = None


class ArtistAnchorRequest(BaseModel):
    # 공연의 어느 아티스트(concert.artist_name 중 하나)를 어느 iTunes 아티스트로 확정할지
    artist: str
    itunes_artist_id: str

File: Prontend/lib/models/setlist.dart
import 'package:flutter/foundation.dart';

/// 곡 한 곡. 백엔드 `schemas/setlist.py`의 `SongEntry`와 대응.
@immutable
class SongEntry {
  final String name;
  final bool encore;

  // [백엔드 수정]
  /// 페스티벌처럼 아티스트가 여럿인 공연에서, 이 곡이 누구 소속인지.
  /// 단독 공연이면 null.
  final String? artist;

  // [백엔드 수정] source 추가 - 'representative'면 대표곡.
  final String? source;

  const SongEntry({
    required this.name,
    this.encore = false,
    this.artist,
    this.source,
  });

  bool get isRepresentative => source == 'representative';

  factory SongEntry.fromJson(Map<String, dynamic> json) {
    return SongEntry(
      name: json['name'] as String,
      encore: json['encore'] as bool? ?? false,
      artist: json['artist'] as String?,
      source: json['source'] as String?,
    );
  }

  // [백엔드 수정] 셋리스트 유저 수정(PATCH) 요청 body 직렬화용.
  Map<String, dynamic> toJson() => {
    'name': name,
    'encore': encore,
    'artist': artist,
    'source': source,
  };
}

/// `GET /concerts/{concertId}/setlist` 응답(실제 셋리스트).
/// 백엔드 `RealSetlistResponse`와 대응.
@immutable
class RealSetlistResponse {
  /// row가 없으면(Setlist.fm 매칭 실패 등) null.
  final String? id;
  final String concertId;
  final String? setlistfmId;
  final List<SongEntry> songs;
  final bool isUserEdited;
  final String? editedUserNickname;

  /// 콘서트에 등록된 아티스트 전원. [songs]에 없는 아티스트도 포함될 수 있음
  /// (placeholder 표시용).
  final List<String> artistNames;

  const RealSetlistResponse({
    this.id,
    required this.concertId,
    this.setlistfmId,
    required this.songs,
    required this.isUserEdited,
    this.editedUserNickname,
    this.artistNames = const [],
  });

  factory RealSetlistResponse.fromJson(Map<String, dynamic> json) {
    return RealSetlistResponse(
      id: json['id'] as String?,
      concertId: json['concert_id'] as String,
      setlistfmId: json['setlistfm_id'] as String?,
      songs: (json['songs'] as List<dynamic>? ?? const [])
          .map((e) => SongEntry.fromJson(e as Map<String, dynamic>))
          .toList(),
      isUserEdited: json['is_user_edited'] as bool,
      editedUserNickname: json['edited_user_nickname'] as String?,
      artistNames: (json['artist_names'] as List<dynamic>? ?? const [])
          .map((e) => e as String)
          .toList(),
    );
  }
}

/// `GET /concerts/{concertId}/setlist/pre` 응답(예상 셋리스트).
/// 백엔드 `PreSetlistResponse`와 대응.
@immutable
class PreSetlistResponse {
  // [백엔드 수정] 예상 셋리가 아직 없으면 null(빈 songs + artistNames만 옴).
  final String? id;
  final String concertId;
  final String? setlistfmId;
  final List<SongEntry> songs;
  final bool isUserEdited;
  final String? editedUserNickname;

  /// 콘서트에 등록된 아티스트(그 날짜 배정이 있으면 그 아티스트로 좁혀짐).
  /// [RealSetlistResponse.artistNames]와 동일 - 단독 공연에서 song.artist가
  /// 비어있는 곡을 음악앱 검색으로 연결할 때 폴백 아티스트로 씀.
  final List<String> artistNames;

  const PreSetlistResponse({
    required this.id,
    required this.concertId,
    this.setlistfmId,
    required this.songs,
    required this.isUserEdited,
    this.editedUserNickname,
    this.artistNames = const [],
  });

  factory PreSetlistResponse.fromJson(Map<String, dynamic> json) {
    return PreSetlistResponse(
      id: json['id'] as String?,
      concertId: json['concert_id'] as String,
      setlistfmId: json['setlistfm_id'] as String?,
      songs: (json['songs'] as List<dynamic>? ?? const [])
          .map((e) => SongEntry.fromJson(e as Map<String, dynamic>))
          .toList(),
      isUserEdited: json['is_user_edited'] as bool,
      editedUserNickname: json['edited_user_nickname'] as String?,
      artistNames: (json['artist_names'] as List<dynamic>? ?? const [])
          .map((e) => e as String)
          .toList(),
    );
  }
}

// [백엔드 수정] 예상 셋리 앵커 후보(공연 아티스트 이름으로 찾은 iTunes 아티스트) 신규.
/// `GET /tickets/{ticketId}/setlist/pre/artist-candidates` 응답 한 건.
/// 동명이인은 [genre]/[topSongs]로 구분.
@immutable
class ArtistCandidate {
  final String itunesArtistId;
  final String artistName;
  final String? genre;
  final List<String> topSongs;
  final String? artworkUrl;

  const ArtistCandidate({
    required this.itunesArtistId,
    required this.artistName,
    this.genre,
    this.topSongs = const [],
    this.artworkUrl,
  });

  factory ArtistCandidate.fromJson(Map<String, dynamic> json) {
    return ArtistCandidate(
      itunesArtistId: json['itunes_artist_id'] as String,
      artistName: json['artist_name'] as String,
      genre: json['genre'] as String?,
      topSongs: (json['top_songs'] as List<dynamic>? ?? const [])
          .map((e) => e as String)
          .toList(),
      artworkUrl: json['artwork_url'] as String?,
    );
  }
}

// [백엔드 수정] 예상 셋리 앵커 후보(iTunes 곡 검색 결과) 신규.
/// `GET /tickets/{ticketId}/setlist/pre/anchor-candidates` 응답 한 건.
/// 고르면 이 곡의 아티스트([itunesArtistId])로 확정됨.
@immutable
class ArtistAnchorCandidate {
  final String itunesArtistId;
  final String artistName;
  final String trackName;
  final String? albumName;
  final String? artworkUrl;

  const ArtistAnchorCandidate({
    required this.itunesArtistId,
    required this.artistName,
    required this.trackName,
    this.albumName,
    this.artworkUrl,
  });

  factory ArtistAnchorCandidate.fromJson(Map<String, dynamic> json) {
    return ArtistAnchorCandidate(
      itunesArtistId: json['itunes_artist_id'] as String,
      artistName: json['artist_name'] as String,
      trackName: json['track_name'] as String,
      albumName: json['album_name'] as String?,
      artworkUrl: json['artwork_url'] as String?,
    );
  }
}

// [백엔드 수정] 공연별 아티스트 연결 수정 후보 신규.
/// `GET /tickets/{ticketId}/artist-identity/candidates` 후보 한 건.
/// [canonicalId]가 없으면 MusicBrainz에만 있는 아티스트(고르면 [mbid]로 연결).
@immutable
class IdentityCandidate {
  final String? canonicalId;
  final String? mbid;
  final String name;
  final String? imageUrl;
  final String? country;
  final String? type;
  final String? disambiguation;
  final String? beginYear;
  // 알아보기용 곡 몇 개(인기순).
  final List<String> topSongs;
  final bool isCurrent;

  const IdentityCandidate({
    this.canonicalId,
    this.mbid,
    required this.name,
    this.imageUrl,
    this.country,
    this.type,
    this.disambiguation,
    this.beginYear,
    this.topSongs = const [],
    this.isCurrent = false,
  });

  factory IdentityCandidate.fromJson(Map<String, dynamic> json) {
    return IdentityCandidate(
      canonicalId: json['canonical_id'] as String?,
      mbid: json['mbid'] as String?,
      name: json['name'] as String,
      imageUrl: json['image_url'] as String?,
      country: json['country'] as String?,
      type: json['type'] as String?,
      disambiguation: json['disambiguation'] as String?,
      beginYear: json['begin_year']?.toString(),
      topSongs: (json['top_songs'] as List<dynamic>? ?? const [])
          .map((e) => e as String)
          .toList(),
      isCurrent: json['is_current'] as bool? ?? false,
    );
  }
}

/// 연결 수정 후보 응답 - [noArtist]면 지금 "연결할 아티스트 없음" 상태.
@immutable
class IdentityCandidatesResponse {
  final bool noArtist;
  final List<IdentityCandidate> candidates;

  const IdentityCandidatesResponse({
    this.noArtist = false,
    this.candidates = const [],
  });

  factory IdentityCandidatesResponse.fromJson(Map<String, dynamic> json) {
    return IdentityCandidatesResponse(
      noArtist: json['no_artist'] as bool? ?? false,
      candidates: (json['candidates'] as List<dynamic>? ?? const [])
          .map((e) => IdentityCandidate.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}
